homeassistant/components/calendar/__init__.py

```diff
@@ -55,70 +45,106 @@ def get_date(date):
 
-    def offset_reached(self):
-        """Have we reached the offset time specified in the event title."""
-        if self._cal_data['start'] is None or \
-           self._cal_data['offset_time'] == dt.dt.timedelta():
-            return False
+def calculate_offset(event, offset):
+    """Calculate event offset.
 
-        return self._cal_data['start'] + self._cal_data['offset_time'] <= \
-            dt.now(self._cal_data['start'].tzinfo)
+    Return the updated event with the offset_time included.
+    """
+    summary = event.get('summary', '')
+    # check if we have an offset tag in the message
+    # time is HH:MM or MM
+    reg = '{}([+-]?[0-9]{{0,2}}(:[0-9]{{0,2}})?)'.format(offset)
+    search = re.search(reg, summary)
+    if search and search.group(1):
+        time = search.group(1)
+        if ':' not in time:
+            if time[0] == '+' or time[0] == '-':
+                time = '{}0:{}'.format(time[0], time[1:])
+            else:
+                time = '0:{}'.format(time)
+
+        offset_time = time_period_str(time)
+        summary = (
+            summary[:search.start()] + summary[search.end():]).strip()
+        event['summary'] = summary
+    else:
+        offset_time = dt.dt.timedelta()  # default it
+
+    event['offset_time'] = offset_time
+    return event
+
+
+def is_offset_reached(event):
+    """Have we reached the offset time specified in the event title."""
+    start = get_date(event['start'])
+    if start is None or event['offset_time'] == dt.dt.timedelta():
+        return False
+
+    return start + event['offset_time'] <= dt.now(start.tzinfo)
+
+
+class CalendarEventDevice(Entity):
+    """A calendar event device."""
 
     @property
-    def name(self):
-        """Return the name of the entity."""
-        return self._name
+    def event(self):
+        """Return the next upcoming event."""
+        raise NotImplementedError()
 
     @property
-    def device_state_attributes(self):
-        """Return the device state attributes."""
-        start = self._cal_data.get('start', None)
-        end = self._cal_data.get('end', None)
-        start = start.strftime(DATE_STR_FORMAT) if start is not None else None
-        end = end.strftime(DATE_STR_FORMAT) if end is not None else None
+    def state_attributes(self):
+        """Return the entity state attributes."""
+        event = self.event
+        if event is None:
+            return None
 
+        event = normalize_event(event)
         return {
-            'message': self._cal_data.get('message', ''),
-            'all_day': self._cal_data.get('all_day', False),
-            'offset_reached': self.offset_reached(),
-            'start_time': start,
-            'end_time': end,
-            'location': self._cal_data.get('location', None),
-            'description': self._cal_data.get('description', None),
+            'message': event['message'],
+            'all_day': event['all_day'],
+            'start_time': event['start'],
```

Comment:
I am not happy with the normalization, however we should probably not make that change in this PR.

There are two types of events (per RFC5545): all day and non-all day. All day events only have a start and end date, so that they are not impacted by timezones. Non-all day have start and end datetimes, so that changing timezone can adjust the event. It looks like we convert all day events to start/end times.

Reply:
Yes, all start and end are converted datetime. We have the boolean `'all_day'` set to signal all day event.

I was thinking about making a data class, eg with attrs, with specified attributes, where an instance of this class should be returned as the event data, instead of having platforms create a dict with the event data. Then we can better control what the event data should be.

But I decided not to do that in this PR, but save it for later.

Reply:
That is a great idea. And yeah, future PR 👍